### PR TITLE
Implement a new filter that denied/allows projects based on total filesize.

### PR DESCRIPTION
**Pull request by electricworry**
WIP. Functionally works, but need to document, add tests, and make some improvements. Also fixed some issues spotted on the way.

### CODE REVIEW

**Review by cooperlees — COMMENTED** [review]
Good to see someone else makes consistent spelling mistakes in code like me (or was it me?)

Looks good, but I am sure you plan it, but please:
- add a quick test over the filter() function at minimum

For some bonus points:
- Please add CHANGES entry
- Please add config + explanation of what it does to documentation

**Comment by electricworry** [comment]
100%. Documentation incoming in just a moment, tests a little later.
I also want to implement human readable size values, e.g. "1G" instead of "1000000000". Do you know of a good quality package that does this conversion already? (from human to int.)

**Comment by cooperlees** [comment]
> Do you know of a good quality package that does this conversion already? (from human to int.)
Maybe https://pypi.org/project/humanfriendly/

File: src/bandersnatch/tests/plugins/test_metadata_plugins.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    def setUp(self) -> None:
+        self.cwd = os.getcwd()
+        self.tempdir = TemporaryDirectory()
+        os.chdir(self.tempdir.name)
+
+    def tearDown(self) -> None:
+        if self.tempdir:
+            assert self.cwd
+            os.chdir(self.cwd)
+            self.tempdir.cleanup()
```
Comment (cooperlees) [inline src/bandersnatch/tests/plugins/test_metadata_plugins.py:30]: Do your tests actually write out any files? I never understood if these are actually needed. I'm guessing you just copy pasted from other test, but was just wondering. I guess it will write out some base mirror directory structure.

File: src/bandersnatch_filter_plugins/allowlist_name.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def filter(self, metadata: Dict) -> bool:
 
     def check_match(self, **kwargs: Any) -> bool:
         """
-        Check if the package name matches against a project that is blocklisted
```
Comment (cooperlees) [inline src/bandersnatch_filter_plugins/allowlist_name.py:67]: Nice catch.

File: src/bandersnatch_filter_plugins/metadata_filter.py [hunk 3]
```diff
@@ -176,14 +179,83 @@ def filter(self, metadata: Dict) -> bool:
...
+                )
+                return
+            if self.max_package_size > 0:
+                if not self.allowlist_package_names:
+                    self.allowlist_package_names = (
+                        self._determine_unfiltered_package_names()
+                    )
+
+                logger.info(
+                    f"Initialized metadata plugin {self.name} to block projects >{self.max_package_size}b"  # noqa: E501
```
Comment (cooperlees) [inline src/bandersnatch_filter_plugins/metadata_filter.py:223]: I guess the E501 is to keep your local editor happy?

File: src/bandersnatch/mirror.py [hunk 4]
```diff
@@ -211,7 +211,6 @@ def __init__(
         self.homedir = self.storage_backend.PATH_BACKEND(str(homedir))
         self.lockfile_path = self.homedir / ".lock"
         self.master = master
-        self.filters = LoadedFilters(load_all=True)
```
Comment (cooperlees) [inline src/bandersnatch/mirror.py:214]: This seems a big change. Why? Can you explain how the functionality has been retained from removing this please or how it's was not needed.

**Review by cooperlees — CHANGES_REQUESTED** [review]
Looks great. Just one question about the change to mirror.py - Just want an explanation to understand and make sure it's not breaking anyting.
- CI does agree the `allowlist_project` plugin is working tho as it uses it ...

File: src/bandersnatch/tests/plugins/test_metadata_plugins.py (hunk 1 above)
Comment (electricworry) [reply src/bandersnatch/tests/plugins/test_metadata_plugins.py:30]: it's true. I'm guilty of copy/pasting. I just presumed that it was a required practice I should follow. :)

File: src/bandersnatch/mirror.py (hunk 4 above)
Comment (electricworry) [reply src/bandersnatch/mirror.py:214]: I should probably not have done this as part of an enhancement, but it's a superflous line. It's set in the superclass's constructor.

File: src/bandersnatch_filter_plugins/metadata_filter.py (hunk 3 above)
Comment (electricworry) [reply src/bandersnatch_filter_plugins/metadata_filter.py:223]: Just a line too long. I'll split it if desired, but I thought it might get a bit confusing with the ternary if-else following.

File: src/bandersnatch/tests/plugins/test_metadata_plugins.py (hunk 1 above)
Comment (ichard26) [reply src/bandersnatch/tests/plugins/test_metadata_plugins.py:30]: And [to some](url), not best practice-y enough since `tearDown()` won't be called if `setUp()` fails.

Comment (cooperlees) [reply src/bandersnatch/tests/plugins/test_metadata_plugins.py:30]: pytest is lots of voodoo. Maybe we can make an issue to test / clean these up.

**Review by cooperlees — APPROVED** [review]
All makes sense to me now. Thanks!
